File: src/features/trading/utils/sales-order-packaging-selection.ts
import {
  type SalesOrderLine,
  type SalesOrderLinePackagingSelection,
  type SalesOrderLinePackagingSelectionSource,
} from '../data/schema'
import { type PackagingProfile } from '@/features/logistics-config/packaging-rules-service'

function matchesProduct(profile: PackagingProfile, productId?: string) {
  if (!productId) {
    return false
  }

  return profile.targets.some(
    (target) => target.entityType === 'product' && target.entityId === productId
  )
}

function dedupeProfiles(profiles: PackagingProfile[]) {
  const seen = new Set<string>()
  return profiles.filter((profile) => {
    if (seen.has(profile.id)) {
      return false
    }

    seen.add(profile.id)
    return true
  })
}

export function getPackagingProfilesForProduct(
  profiles: PackagingProfile[],
  productId?: string,
  activeOnly = false
): PackagingProfile[] {
  return dedupeProfiles(
    profiles.filter(
      (profile) => matchesProduct(profile, productId) && (!activeOnly || profile.isActive)
    )
  )
}

export function buildSalesOrderLinePackagingSelection(
  profile: PackagingProfile,
  source: SalesOrderLinePackagingSelectionSource
): SalesOrderLinePackagingSelection {
  return {
    profileId: profile.id,
    profileCode: profile.code,
    profileName: profile.name,
    packagingType: profile.packagingType,
    length: profile.length,
    width: profile.width,
    height: profile.height,
    dimensionUnitCode: profile.dimensionUnitCode,
    netWeight: profile.netWeight,
    grossWeight: profile.grossWeight,
    weightUnitCode: profile.weightUnitCode,
    capacity: profile.capacity,
    capacityUnitCode: profile.capacityUnitCode,
    source,
  }
}

export function findPackagingProfileById(
  profiles: PackagingProfile[],
  profileId?: string
): PackagingProfile | undefined {
  if (!profileId) {
    return undefined
  }

  return profiles.find((profile) => profile.id === profileId)
}

export function resolveAutoPackagingProfile(
  profiles: PackagingProfile[],
  productId?: string
): PackagingProfile | undefined {
  const activeProfiles = getPackagingProfilesForProduct(profiles, productId, true)
  if (activeProfiles.length === 0) {
    return undefined
  }

  const defaultProfiles = dedupeProfiles(
    activeProfiles.filter((profile) =>
      profile.targets.some(
        (target) =>
          target.entityType === 'product' &&
          target.entityId === productId &&
          target.isDefault
      )
    )
  )

  if (defaultProfiles.length === 1) {
    return defaultProfiles[0]
  }

  if (activeProfiles.length === 1) {
    return activeProfiles[0]
  }

  return undefined
}

export function resolveSalesOrderLinePackagingSelection(
  line: SalesOrderLine,
  profiles: PackagingProfile[]
): SalesOrderLinePackagingSelection | undefined {
  if (line.selectedPackaging?.profileId) {
    return line.selectedPackaging
  }

  const autoProfile = resolveAutoPackagingProfile(profiles, line.productId)
  if (!autoProfile) {
    return undefined
  }

  return buildSalesOrderLinePackagingSelection(autoProfile, 'auto')
}

export function createPackagingProfileFromSelection(
  selection: SalesOrderLinePackagingSelection
): PackagingProfile {
  return {
    id: selection.profileId,
    code: selection.profileCode,
    name: selection.profileName,
    packagingType: selection.packagingType,
    length: selection.length,
    width: selection.width,
    height: selection.height,
    dimensionUnitCode: selection.dimensionUnitCode,
    netWeight: selection.netWeight,
    grossWeight: selection.grossWeight,
    weightUnitCode: selection.weightUnitCode,
    capacity: selection.capacity,
    capacityUnitCode: selection.capacityUnitCode,
    assemblySource: '',
    isActive: true,
    notes: '',
    targets: [],
  }
}

File: server/services/sales_order_packaging_selection.go
package services

import (
	"encoding/json"
	"strings"
	"xdfc-server/models"

	"gorm.io/gorm"
)

type salesOrderPackagingProfileCandidate struct {
	Profile   models.PackagingProfile
	IsDefault bool
}

func normalizeSalesOrderLinePackagingSelectionSource(value string) string {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "auto":
		return "auto"
	default:
		return "manual"
	}
}

func encodeSalesOrderLinePackagingSelection(selection *SalesOrderLinePackagingSelectionPayload) json.RawMessage {
	if selection == nil || strings.TrimSpace(selection.ProfileID) == "" {
		return nil
	}

	normalized := *selection
	normalized.Source = normalizeSalesOrderLinePackagingSelectionSource(normalized.Source)
	payload, err := json.Marshal(normalized)
	if err != nil {
		return nil
	}
	return payload
}

func decodeSalesOrderLinePackagingSelection(raw json.RawMessage) *SalesOrderLinePackagingSelectionPayload {
	if len(raw) == 0 || string(raw) == "null" {
		return nil
	}

	var selection SalesOrderLinePackagingSelectionPayload
	if err := json.Unmarshal(raw, &selection); err != nil {
		return nil
	}
	if strings.TrimSpace(selection.ProfileID) == "" {
		return nil
	}
	selection.Source = normalizeSalesOrderLinePackagingSelectionSource(selection.Source)
	return &selection
}

func buildSalesOrderLinePackagingSelectionFromProfile(
	profile models.PackagingProfile,
	source string,
) *SalesOrderLinePackagingSelectionPayload {
	return &SalesOrderLinePackagingSelectionPayload{
		ProfileID:         profile.ID,
		ProfileCode:       profile.Code,
		ProfileName:       profile.Name,
		PackagingType:     profile.PackagingType,
		Length:            profile.Length,
		Width:             profile.Width,
		Height:            profile.Height,
		DimensionUnitCode: profile.DimensionUnitCode,
		NetWeight:         profile.NetWeight,
		GrossWeight:       profile.GrossWeight,
		WeightUnitCode:    profile.WeightUnitCode,
		Capacity:          profile.Capacity,
		CapacityUnitCode:  profile.CapacityUnitCode,
		Source:            normalizeSalesOrderLinePackagingSelectionSource(source),
	}
}

func collapseSalesOrderPackagingCandidates(
	candidates []salesOrderPackagingProfileCandidate,
) []salesOrderPackagingProfileCandidate {
	if len(candidates) == 0 {
		return nil
	}

	byProfileID := make(map[string]salesOrderPackagingProfileCandidate, len(candidates))
	orderedProfileIDs := make([]string, 0, len(candidates))
	for _, candidate := range candidates {
		profileID := strings.TrimSpace(candidate.Profile.ID)
		if profileID == "" {
			continue
		}

		existing, ok := byProfileID[profileID]
		if !ok {
			byProfileID[profileID] = candidate
			orderedProfileIDs = append(orderedProfileIDs, profileID)
			continue
		}

		if candidate.IsDefault {
			existing.IsDefault = true
			byProfileID[profileID] = existing
		}
	}

	result := make([]salesOrderPackagingProfileCandidate, 0, len(orderedProfileIDs))
	for _, profileID := range orderedProfileIDs {
		result = append(result, byProfileID[profileID])
	}
	return result
}

func findSalesOrderPackagingCandidateByProfileID(
	candidates []salesOrderPackagingProfileCandidate,
	profileID string,
) *salesOrderPackagingProfileCandidate {
	trimmedProfileID := strings.TrimSpace(profileID)
	if trimmedProfileID == "" {
		return nil
	}

	for _, candidate := range collapseSalesOrderPackagingCandidates(candidates) {
		if candidate.Profile.ID == trimmedProfileID {
			candidateCopy := candidate
			return &candidateCopy
		}
	}
	return nil
}

func resolveAutoSalesOrderLinePackagingSelection(
	candidates []salesOrderPackagingProfileCandidate,
) *SalesOrderLinePackagingSelectionPayload {
	collapsedCandidates := collapseSalesOrderPackagingCandidates(candidates)
	if len(collapsedCandidates) == 0 {
		return nil
	}

	defaultCandidates := make([]salesOrderPackagingProfileCandidate, 0, len(collapsedCandidates))
	for _, candidate := range collapsedCandidates {
		if candidate.IsDefault {
			defaultCandidates = append(defaultCandidates, candidate)
		}
	}

	if len(defaultCandidates) == 1 {
		return buildSalesOrderLinePackagingSelectionFromProfile(defaultCandidates[0].Profile, "auto")
	}
	if len(collapsedCandidates) == 1 {
		return buildSalesOrderLinePackagingSelectionFromProfile(collapsedCandidates[0].Profile, "auto")
	}
	return nil
}

func loadSalesOrderPackagingCandidatesByProductID(
	tx *gorm.DB,
	lines []models.SalesOrderLine,
) (map[string][]salesOrderPackagingProfileCandidate, error) {
	productIDs := make([]string, 0, len(lines))
	seenProductIDs := make(map[string]struct{}, len(lines))
	for _, line := range lines {
		productID := strings.TrimSpace(line.ProductID)
		if productID == "" {
			continue
		}
		if _, exists := seenProductIDs[productID]; exists {
			continue
		}
		seenProductIDs[productID] = struct{}{}
		productIDs = append(productIDs, productID)
	}

	if len(productIDs) == 0 {
		return map[string][]salesOrderPackagingProfileCandidate{}, nil
	}

	var targets []models.PackagingProfileTarget
	if err := tx.
		Where("LOWER(entity_type) = ?", "product").
		Where("entity_id IN ?", productIDs).
		Find(&targets).Error; err != nil {
		return nil, err
	}
	if len(targets) == 0 {
		return map[string][]salesOrderPackagingProfileCandidate{}, nil
	}

	profileIDs := make([]string, 0, len(targets))
	seenProfileIDs := make(map[string]struct{}, len(targets))
	for _, target := range targets {
		profileID := strings.TrimSpace(target.PackagingProfileID)
		if profileID == "" {
			continue
		}
		if _, exists := seenProfileIDs[profileID]; exists {
			continue
		}
		seenProfileIDs[profileID] = struct{}{}
		profileIDs = append(profileIDs, profileID)
	}

	var profiles []models.PackagingProfile
	if err := tx.
		Where("id IN ?", profileIDs).
		Where("is_active = ?", true).
		Find(&profiles).Error; err != nil {
		return nil, err
	}

	profileByID := make(map[string]models.PackagingProfile, len(profiles))
	for _, profile := range profiles {
		profileByID[strings.TrimSpace(profile.ID)] = profile
	}

	result := make(map[string][]salesOrderPackagingProfileCandidate, len(productIDs))
	for _, target := range targets {
		profile, ok := profileByID[strings.TrimSpace(target.PackagingProfileID)]
		if !ok {
			continue
		}
		productID := strings.TrimSpace(target.EntityID)
		result[productID] = append(result[productID], salesOrderPackagingProfileCandidate{
			Profile:   profile,
			IsDefault: target.IsDefault,
		})
	}

	return result, nil
}

func normalizeSalesOrderLinePackagingSelectionsTx(
	tx *gorm.DB,
	previousLines []models.SalesOrderLine,
	nextLines []models.SalesOrderLine,
) error {
	candidatesByProductID, err := loadSalesOrderPackagingCandidatesByProductID(tx, nextLines)
	if err != nil {
		return err
	}

	previousLineByNo := make(map[int]models.SalesOrderLine, len(previousLines))
	for _, line := range previousLines {
		previousLineByNo[line.LineNo] = line
	}

	for index := range nextLines {
		currentLine := &nextLines[index]
		productID := strings.TrimSpace(currentLine.ProductID)
		if productID == "" {
			currentLine.SelectedPackaging = nil
			continue
		}

		currentSelection := decodeSalesOrderLinePackagingSelection(currentLine.SelectedPackaging)
		previousLine, hasPrevious := previousLineByNo[currentLine.LineNo]
		previousSelection := decodeSalesOrderLinePackagingSelection(previousLine.SelectedPackaging)
		productChanged := !hasPrevious || strings.TrimSpace(previousLine.ProductID) != productID
		candidates := candidatesByProductID[productID]

		if hasPrevious && !productChanged {
			if currentSelection != nil {
				if previousSelection != nil && currentSelection.ProfileID == previousSelection.ProfileID {
					currentLine.SelectedPackaging = encodeSalesOrderLinePackagingSelection(currentSelection)
					continue
				}
				if candidate := findSalesOrderPackagingCandidateByProfileID(candidates, currentSelection.ProfileID); candidate != nil {
					currentLine.SelectedPackaging = encodeSalesOrderLinePackagingSelection(
						buildSalesOrderLinePackagingSelectionFromProfile(candidate.Profile, currentSelection.Source),
					)
					continue
				}
			}
			if previousSelection != nil {
				currentLine.SelectedPackaging = encodeSalesOrderLinePackagingSelection(previousSelection)
				continue
			}
		}

		if currentSelection != nil {
			if candidate := findSalesOrderPackagingCandidateByProfileID(candidates, currentSelection.ProfileID); candidate != nil {
				currentLine.SelectedPackaging = encodeSalesOrderLinePackagingSelection(
					buildSalesOrderLinePackagingSelectionFromProfile(candidate.Profile, currentSelection.Source),
				)
				continue
			}
		}

		currentLine.SelectedPackaging = encodeSalesOrderLinePackagingSelection(
			resolveAutoSalesOrderLinePackagingSelection(candidates),
		)
	}

	return nil
}
